constructor/winexe.py

```diff
@@ -87,6 +87,8 @@ def setup_envs_commands(info, dir_path):
             "lockfile_txt_abspath": join(dir_path, "conda-meta", "initial-state.explicit.txt"),
             "conda_meta": r"$INSTDIR\conda-meta",
             "history_abspath": join(dir_path, "conda-meta", "history"),
+            "frozen_abspath": join(dir_path, "conda-meta", "frozen"),
+            "has_frozen": bool(info.get("freeze_base")),
```

Comment:
```suggestion
            "frozen_abspath": join(dir_path, "conda-meta", "frozen") if info.get("freeze_base", {}).get("conda") is not None else "",
```

This is more compact and doesn't add redundant `has_*` template variables. Also, you need to be specific about the `conda` subkey here.

Reply:
I agree your suggestion is more compact, but it's also a little harder to read. I went with the redundancy for readability and to match the existing code. I do agree about the subkey, so how's this?
```
"frozen_abspath": join(dir_path, "conda-meta", "frozen"),
"has_frozen": bool(info.get("freeze_base", {}).get("conda")),
```

Reply:
I would do the same thing for the extra envs as well.

Reply:
This will be incorrect because `bool({})` evaluates to `False`, but we need it to be `True`. You don't have to store this as a template variable. You could just create

```
has_frozen_conda = info.get("freeze_base", {}).get("conda") is not None
```

and then later add the following to your template

```
"frozen_abspath": join(dir_path, "conda-meta", "frozen") if has_frozen else ""
```

Reply:
I almost forgot about that. My original code handles that because `bool(info.get("freeze_base"))` is `True if "conda" in info.get("freeze_base")` and `preconda.py` will return `if "conda" not in freeze_info`, so I don't think it's necessary to do that check again here. Once again, the original code is also more readable this way and future proof.

Reply:
`preconda` returns if `"conda"` is not in `freeze_info` *at the moment*. It won't be once we add new package managers like `pip`. Once new package managers are added, you will also have to update the code here - which will make it less future-proof because that's easily forgotten.

As for legibility, I'm not sure I follow which part is illegible. You have to run define `has_frozen_conda` somewhere and `frozen_abspath` is a standard ternary operator (as far as Python has one). Your solution adds an additional template variable though, mine does not.

Reply:
In the end, if you think that adding more template variables helps, I won't block this. But I strongly suggest that the `has_frozen_conda` check is made more robust.

Reply:
We wouldn't have to update the code here if we leave as is because `bool(info.get("freeze_base"))` would still be `True` if/and we would only need to update `write_frozen` in `preconda` to include the additional package managers.

Reply:
Exactly, it would still be `True` even if `freeze_base` is:

```
freeze_base:
  pip: {}
```

So, you'd be writing `conda-meta/frozen` even if that's not what the `construct.yaml` file says.

Reply:
If that's not desired then we could add a helper function or simply check if the file exists (i.e. remove `frozen_abspath` and change `has_frozen` to `isfile(join(dir_path, "conda-meta", "frozen"))`. Otherwise, we would have to add something like `info.get("freeze_base", {}).get("pip")` to your suggested `info.get("freeze_base", {}).get("conda")` when the time comes. Honestly, the more I think about it, if we add additional package managers and they don't have the exact same end goal as `conda` of creating a frozen marker file, then we're going to have to do updates in multiple places anyway. Originally, I was thinking even if `pip` had its own rules, it could all be addressed in `write_frozen` but the end goal to have `conda-meta/frozen` would remain the same regardless of the content of the file or if the file was renamed to `frozen`, etc.

Reply:
> Honestly, the more I think about it, if we add additional package managers and they don't have the exact same end goal as conda of creating a frozen marker file, then we're going to have to do updates in multiple places anyway.

They do rely on marker files, but not the same as `conda`. `conda-meta/frozen` is only respected by `conda`. `pip` requires a different file in a different format (and unfortunately, in a different location for Windows and Unix): https://packaging.python.org/en/latest/specifications/externally-managed-environments/#marking-an-interpreter-as-using-an-external-package-manager

Yes, we have to *add* to multiple places to add `pip`, but we don't have to change pre-existing code.

> (i.e. remove frozen_abspath and change has_frozen to isfile(join(dir_path, "conda-meta", "frozen"))

You unfortunately do need `frozen_abspath` because you need to write the absolute path of the file on the build system into the NSIS file (not the destination path, the source path). That's what's telling NSIS to package that file and extract it into the directory set by `SetOutPath`. What you don't necessarily need is `has_frozen` because that could be replaced with:

```
{%- if env.frozen_abspath` %}
  File {{ env.frozen_abspath }}
{% endif %}
```

> or simply check if the file exists 

That's another option, and I believe this what you did for SH installers, too. I don't know off the top of my head if `write_files` is called before the template variables are populated though - if they are, that's a very robust approach. If they're not, you can move the function call to before the variables are populated in the dictionary.